### PR TITLE
Resolve variadic parameters.

### PR DESCRIPTION
**Pull request by serge-kvashnin**
`ParameterResolver::resolveParameters()` was iterating through the reflection parameters and fetching appropriate values from either named parameters or definition parameters. This meant the number of resolved arguments matched the number of reflection parameters, even if a parameter was variadic. This change adds a check for variadic parameters and adds the rest of the parameters to the `args` array accordingly.

### CODE REVIEW

**Comment by serge-kvashnin** [comment]
Solves https://github.com/PHP-DI/PHP-DI/issues/881

**Comment by mnapoli** [comment]
Any significant feature addition like this requires a new integration test, can you add it? I'm not sure how it behaves with the compiled container.

**Comment by serge-kvashnin** [comment]
@mnapoli You're right. Added test and handling of variadic parameter for `ObjectCreationCompiler`.

File: .php-cs-fixer.php [hunk 1]
```diff
@@ -58,6 +58,7 @@
     'blank_line_between_import_groups' => false,
     'global_namespace_import' => false,
     'nullable_type_declaration_for_default_null_value' => false,
+    'nullable_type_declaration' => false,
```
Comment (mnapoli) [inline .php-cs-fixer.php:61]: Why this change?

File: tests/IntegrationTest/CompiledContainerTest.php [hunk 2]
```diff
@@ -177,6 +177,41 @@ public function proxy_classes_can_be_pregenerated_at_compile_time()
         $this->assertEquals(2, $countProxyClasses);
     }
 
+    /**
+     * @test
+     * @see https://github.com/PHP-DI/PHP-DI/issues/881
+     */
+    public function works_with_variadic_parameters()
```
Comment (mnapoli) [inline tests/IntegrationTest/CompiledContainerTest.php:184]: Why not add this to the base containter test class? This isn't specific to the compiled container, variadic parametrs should be supported (and so tested) on both compiled and non-compiled containers.

Comment (serge-kvashnin) [reply tests/IntegrationTest/CompiledContainerTest.php]: Created separate test case along with other issue cases, for both compiled and non-compiled containers.

File: .php-cs-fixer.php (hunk 1 above)
Comment (mnapoli) [reply .php-cs-fixer.php:61]: Oh you mean that fixes master?

A bit more clarity would help here 🙂

File: tests/IntegrationTest/Issues/Issue881Test.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+use DI\Test\IntegrationTest\Issues\Issue881\A;
+use DI\Test\IntegrationTest\Issues\Issue881\B;
+use DI\Test\IntegrationTest\Issues\Issue881\C;
+use DI\Test\IntegrationTest\Issues\Issue881\ClassWithVariadicParameter;
+use function DI\create;
+use function DI\get;
+
+final class Issue881Test extends BaseContainerTest
+{
+    public function testContainerWithVariadicParameters(): void
```
Comment (mnapoli) [inline tests/IntegrationTest/Issues/Issue881Test.php:18]: Can you move this to `DI\Test\IntegrationTest\Definitions\CreateDefinitionTest` and change the signature like the other test methods to merge both tests in 1:

```suggestion
    /**
     * @dataProvider provideContainer
     */
    public function test_with_variadics(ContainerBuilder $builder)
```

File: tests/IntegrationTest/Issues/Issue881/ClassWithVariadicParameter.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+declare(strict_types=1);
+
+namespace DI\Test\IntegrationTest\Issues\Issue881;
+
+final class ClassWithVariadicParameter
```
Comment (mnapoli) [inline tests/IntegrationTest/Issues/Issue881/ClassWithVariadicParameter.php:7]: This class can be moved to `DI\Test\IntegrationTest\Definitions\ObjectDefinition`

File: .php-cs-fixer.php (hunk 1 above)
Comment (serge-kvashnin) [reply .php-cs-fixer.php:61]: Yes, this change disables the [PhpCsFixer](https://cs.symfony.com/doc/rules/language_construct/nullable_type_declaration.html) `nullable_type_declaration` check, preventing the master build from failing. [Since PHP 8.4](https://wiki.php.net/rfc/deprecate-implicitly-nullable-types) will deprecate this kind of declaration, I suggest creating a separate Issue/PR to address it. You can then decide whether to standardize on union types or the question mark operator (whichever you think is preferable).

File: tests/IntegrationTest/Issues/Issue881/ClassWithVariadicParameter.php (hunk 4 above)
Comment (serge-kvashnin) [reply tests/IntegrationTest/Issues/Issue881/ClassWithVariadicParameter.php]: Done [8854fe6](https://github.com/PHP-DI/PHP-DI/pull/883/commits/8854fe64da554a36d93b3a79cdf5045745fce58b).

File: tests/IntegrationTest/Issues/Issue881Test.php (hunk 3 above)
Comment (serge-kvashnin) [reply tests/IntegrationTest/Issues/Issue881Test.php]: Done [8854fe6](https://github.com/PHP-DI/PHP-DI/pull/883/commits/8854fe64da554a36d93b3a79cdf5045745fce58b).

**Comment by serge-kvashnin** [comment]
@mnapoli Please don't merge it yet. After digging into issues https://github.com/PHP-DI/PHP-DI/issues/619 and https://github.com/PHP-DI/PHP-DI/issues/802, and considering [Ben's comment](https://github.com/PHP-DI/PHP-DI/issues/802#issuecomment-1478713488) I added some edge case tests. Turns out, there are still a few things I need to work through. I'll dive deeper and get this sorted as soon as I can.

**Comment by serge-kvashnin** [comment]
Summary of changes:

1. I introduced `ksort` because for definitions where specific parameters are defined and not ordered, the array might look like `[1 => 'bar', 0 => 'foo']`. This ensures array_slice will properly fetch the rest of the variadic parameters.
2. I also made `isVariadic` check the first priority check so that it works properly when variadic parameters are not passed.
3. I also enhanced the `ReflectionBasedAutowiring::autowire()` method with iteration through defined method injections. This involves completing method injections with parameters taken from reflection but not defined explicitly (was already done for constructor). It has nothing to do with variadic parameters, but I spotted this issue when I was defining only parameter `b` while skipping `a` (which was type-hinted).

**Comment by joachim-n** [comment]
I can't comment on the code in general as I don't know this codebase at all, but I can confirm that this PR fixes #881 for me.
